networkx/algorithms/distance_regular.py

```diff
@@ -146,37 +149,66 @@ def intersection_array(G):
     --------
     global_parameters
     """
+    # the input graph is very unlikely to be distance-regular: here are the
+    # number a(n) of connected simple graphs, and the number b(n) of
+    # distance-regular graphs among them:
+    #
+    #    n  | 1 2 3 4  5   6   7     8      9       10
+    #  -----+------------------------------------------------------------------
+    #  a(n) | 1 1 2 6 21 112 853 11117 261080 11716571 https://oeis.org/A001349
+    #  b(n) | 1 1 1 2  2   4   2     5      4        7 https://oeis.org/A241814
+    #
+    # in light of this, let's compute shortest path lengths as we go instead of
+    # precomputing them all
     # test for regular graph (all degrees must be equal)
-    if len(G) == 0:
-        raise nx.NetworkXPointlessConcept("Graph has no nodes.")
-    degree = iter(G.degree())
-    (_, k) = next(degree)
-    for _, knext in degree:
-        if knext != k:
-            raise nx.NetworkXError("Graph is not distance regular.")
-        k = knext
-    path_length = dict(nx.all_pairs_shortest_path_length(G))
-    diameter = max(max(path_length[n].values()) for n in path_length)
+    if not nx.is_regular(G) or not nx.is_connected(G):
+        raise nx.NetworkXError("Graph is not distance regular.")
```

Comment:
I don't know that it is more intuitive, but I suppose it would be faster (although I haven't checked). However, your suggestion might be safer, because `nx.is_regular(G)` and `nx.is_connected(G)` both crash on a graph with no vertices and no edges (but they behave fine if `G` has vertices but no edges). So I guess we can replace the test with: `if nx.is_empty(G):`, which is indeed more readable.

Reply:
What I meant with my comment was that it could be worth checking whether `G` is the null graph so you can explicitly return the `PointlessConcept` error with a suitable message, rather than relying on this to be done implicitly (with an unclear "Graph has no nodes." message) in `nx.is_regular(G)`.

Using `not nx.is_empty(G)` as check shouldn't work (this only checks for the presence of edges, but the regularity and connectedness conditions are obviously much stronger).

Note that the crashes shouldn't be problematic in and of themselves; the whole function relies on errors to know whether it was successful. My comment was a nit about raising them explicitly being clearer for users.